### PR TITLE
ci coverage: fail PRs on per-package coverage decrease or overall floor breach

### LINKED ISSUE #7799: Surface coverage diff on every PR
Sub-task of epic #7660.

## Problem
Contributors don''t see whether their PR increases or decreases coverage. Coverage trend is tracked in CI but not visible at PR-time, so reviewers can''t enforce "no coverage regression" without manual digging.

## Goal
Every PR shows a coverage delta:
- Overall coverage change (e.g., `+0.3%` or `-1.2%`)
- Per-package coverage change for files touched in the PR
- Flag if any touched file''s coverage drops below a configurable floor

## Approach (to be scoped)
- Use existing `mage coverage:ci` output as the base
- PR comment bot OR GitHub status check OR PR annotation
- Surface in the PR review experience without being noisy

## Out of scope
- Blocking merges on coverage decrease (separate policy decision)
- Net new coverage measurement infrastructure (#7524 builds the foundation)

### PR DESCRIPTION
**Pull request by hemarina**
Closes #7799

## Summary

Replaces the legacy single-threshold gate (`Test-CodeCoverageThreshold.ps1`) with a **fail-loud, two-gate PR coverage check** so coverage regressions block merges instead of silently slipping in.

### Two-gate enforcement (PR builds only)

A PR fails (`exit 2`) if **either** gate breaches:

| Gate | Default | Purpose |
|---|---|---|
| **Per-package decrease** (`-MaxPackageDecrease`) | `0.5` pp | Catches localized regressions in any touched package |
| **Overall floor** (`-MinOverallCoverage`) | `69.0` % | Prevents the codebase as a whole from sliding |

Both are configurable via pipeline parameters; either gate can be disabled independently with the `-1` sentinel. Any non-zero exit other than `2` indicates a script/infra error, not a gate breach.

### New: `eng/scripts/Get-CoverageDiff.ps1`

Replaces `Test-CodeCoverageThreshold.ps1` for CI use. Features:
- Per-package coverage delta scoped to PR-touched packages (via `git diff --diff-filter=AMRD origin/<targetBranch>...HEAD`).
- Falls back to top-N package mode when no changed-files input is supplied.
- Locale-invariant number formatting (no `,` decimal leakage on non en-US agents).
- Graceful baseline-skip when the target branch has no successful prior build (advisory exit 0, with loud `##vso[task.logissue]` annotation so the skip is visible).
- Strict input validation, malformed-profile tolerance, deletion-only PR support.
- Precise exit codes: `0` clean, `2` gate breach (with `-FailOnGate`), other non-zero = script error.

### New: tests + CI wiring

- **`eng/scripts/Get-CoverageDiff.Tests.ps1`** — 37 Pester tests covering per-package scoping, changed-file input handling, profile parsing edge cases, both gates and their boundaries, locale invariance, `-1` disable sentinels, deletion-only PRs, `-TopN`/`-MinDelta` filters, input validation.
- **`cli/azd/magefile_test.go`** — 11 Go tests covering `resolveCoverageFile`, `resolveBaselineFile`, and `resolveChangedFilesForDiff` (incl. real-git happy path with bare-upstream simulation, on-main, detached-HEAD, strict-mode error, **and a concurrency-regression assertion** that two resolves on the same repo produce distinct temp paths).
- Both wired into `cli-ci.yml` as new jobs: `coverage-script-tests` (Pester, pinned to v5.7.1, `permissions: contents: read`) and `magefile-tests` (Go, `permissions: contents: read`).

### Pipeline (`eng/pipelines/templates/stages/code-coverage-upload.yml`)

- Runs the gate **only on PR builds** (main/release/scheduled remain artifact-publish only).
- Resolves baseline by querying the latest successful target-branch build for the same pipeline definition; downloads its `cover-unit` + `cover-int` artifacts.
- Retries baseline lookup on transient ADO REST failures; emits a loud annotation if all retries fail.
- **Validates artifacts are non-empty** (not just present) on both current-build and baseline sides — catches partial-publish legs that would silently shrink the merged coverage total.
- Detects platform-coverage gaps (e.g., a package only covered on Linux) and warns rather than mis-attributing the gap to the PR.
- Surfaces the gate verdict in the build log (PASS/FAIL block listing both gates and any breaches).
- `mage` pinned to **`v1.17.2`** (must stay in sync with `cli/azd/go.mod`; comment in the pipeline reminds future maintainers).

### Mage targets (`cli/azd/magefile.go`)

New `Coverage` namespace with local-developer ergonomics:
- `mage coverage:diff` — preview the gate locally against `origin/main`. Advisory by default; set `COVERAGE_FAIL_ON_DECREASE=1` to make it fail-loud (exit 2).
- `mage coverage:pr` — fail-loud preview that mirrors the CI gate exactly.
- Other helpers for unit/integration coverage merge and report.
- Uses `os.CreateTemp` for the changed-files list so two concurrent local runs can't clobber each other's input and silently produce wrong gate results.
- Stale-merge-dir cleanup errors are checked + wrapped (no silent `os.RemoveAll` failures).

### Docs (`cli/azd/docs/code-coverage-guide.md`)

Rewritten contributor guide: pipeline flow, the two-gate model, mage targets, env vars (with precise `exit 2` semantics), troubleshooting runbook, branch-protection setup (with `gh api` admin commands), and a worked example showing a 1.0pp drop tripping the per-package gate.

### Cleanup

- `Test-CodeCoverageThreshold.ps1` doc-comment refreshed (removed stale `release-cli.yml` reference; clarified it is now local-only via `Get-LocalCoverageReport.ps1`).
- Dead `MinimumCoveragePercent: 65` parameter removed from `release-cli.yml`.
- Removed unused `GH_TOKEN` env var from the PR coverage diff step (no `gh` invocations in that step).
- Pester install hardened: removed `-SkipPublisherCheck`, pinned to `-RequiredVersion 5.7.1`.
- `cspell.yaml` overrides extended for new domain words (`textfmt`, `logissue`, `runbook`, `covcounters`, `covmeta`, `AMRD`).

## Validation

- ✅ `coverage-script-tests` GHA job: **37/37 Pester tests passing**
- ✅ `magefile-tests` GHA job: **11/11 Go tests passing** (incl. concurrency regression check)
- ✅ `cspell lint` clean across all touched files
- ✅ `go vet -tags mage .` clean
- ✅ YAML parses (`code-coverage-upload.yml`, `release-cli.yml`, `cli-ci.yml`)
- ✅ End-to-end ADO pipeline run on this PR — gate stage executes; baseline-skip path verified (advisory) when no successful main build exists for the running pipeline definition

## Effective behavior after merge

| Build type | Behavior |
|---|---|
| **PR** | Two-gate enforcement; fails build (`exit 2`) on per-package decrease > 0.5pp or overall < 69% |
| **main / release / scheduled / manual** | Publish coverage artifacts only — no enforcement |
| **Local (`mage coverage:diff`)** | Advisory preview; opt-in fail-loud via `COVERAGE_FAIL_ON_DECREASE=1` |
| **Local (`mage coverage:pr`)** | Fail-loud preview mirroring CI |

## Follow-ups (out of scope)

- Set the new `coverage-script-tests`, `magefile-tests`, and `azure-dev - cli (CodeCoverage_Upload Upload)` checks as **required** in branch protection (admin-only; `gh api` command in the runbook).
- Centralized coverage-trend dashboard (Jon's ask) — separate work item; this PR establishes the per-PR gate; trend visualization can layer on top of the same artifacts.

### CODE REVIEW

**Review by wbreza — COMMENTED** [review]
## 🔍 Code Review — PR #8092

**ci coverage: fail PRs when touched files drop below per-file coverage floor**

> **General note:** PRs should always include a description of their changes regardless of how trivial they may seem. A good description helps reviewers understand intent, provides context for future readers, and satisfies governance checks.

---

### P0 — Blocks Merge

**1. Empty PR description + missing issue link**
- PR body is empty. No `Fixes #7799` despite the branch name referencing it.
- `pr-governance` CI check **FAILS** because of this — mechanical merge blocker.
- **Fix:** Add PR description with `Fixes #7799` and a summary of the two-gate coverage model.

---

### P1 — Should Fix Before Merge

**2. PR title says "per-file" but implementation is "per-package"**
- Title: *"fail PRs when touched files drop below per-file coverage floor"*
- Reality: Gates operate at **Go package** granularity, not per-file.
- **Fix:** Update title to reflect per-package granularity (e.g., *"ci coverage: add two-gate PR coverage check (per-package regression + absolute floor)"*).

**3. AGENTS.md says 65% floor but code uses 69%**
- AGENTS.md: *"overall coverage falls below 65%"*
- Actual default everywhere else (script, pipeline, guide): **69%**
- **Fix:** Update AGENTS.md to say 69%.

**4. GitHub Actions `coverage-script-tests` job missing `permissions:` block**
- Job inherits `pull-requests: write` from workflow-level but only needs `contents: read`.
- Violates repo's least-privilege convention.
- **Fix:** Add `permissions: { contents: read }` to the job.

**5. `-SkipPublisherCheck` on Pester install (supply chain)**
- `Install-Module -Name Pester ... -SkipPublisherCheck` disables code signing verification.
- **Fix:** Remove `-SkipPublisherCheck` if possible; at minimum pin to `-RequiredVersion 5.x.x`.

**6. `go install mage@latest` — unpinned dependency (supply chain)**
- CI installs `mage@latest` which is non-deterministic across builds.
- **Fix:** Pin to a specific version (e.g., `mage@v1.15.0`) or use the version from go.mod.

**7. `os.RemoveAll` errors silently ignored in magefile.go**
- Lines 536-537: `_ = os.RemoveAll(tmpUnit)` — if removal fails (Windows file locking), stale covdata contaminates the merge.
- **Fix:** Check and return the error.

---

### P2 — Consider

**8. No Go unit tests for new magefile functions** — `resolveChangedFilesForDiff()`, `Coverage.PR()`, `Coverage.Report()` have zero test coverage. Pester suite covers downstream behavior, so lower urgency.

**9. Parameter naming mismatch** — Docs use `PackageCoverageDecreaseTolerance` but code uses `MaxPackageDecrease`. Align terminology for searchability.

**10. Artifact validation only checks directory existence** — `Test-Path` returns true for empty dirs. A failed platform leg with an empty dir would pass validation silently.

**11. Exit code documentation inconsistent** — Some docs say "exit non-zero", code uses exit 2 specifically. Standardize across docs and mage comments.

---

### ✅ Positives

- **Well-designed two-gate model** — per-package regression + absolute floor is a sound approach
- **Excellent Pester test suite** — 857 lines covering boundary conditions, locale-invariant formatting, gate combinations
- **Good fail-closed design** — git failures fall back to full package scan, never silently skip gates
- **`mage coverage:pr`** — great DX for contributors to preview the gate locally
- **Consolidated coverage reporting** — single source of truth via `mage coverage:report`

**Review by vhvb1989 — APPROVED** [review]

File: cli/azd/magefile.go [hunk 1]
```diff
@@ -1159,40 +1229,196 @@ func (Coverage) PR() error {
...
-		"-Tag", "<!-- coverage-diff -->",
-	)
+		return skipOrFail(fmt.Sprintf("git diff failed: %v", err))
+	}
+	diff = strings.TrimSpace(diff)
+	if diff == "" {
+		return skipOrFail("no files changed vs origin/main")
+	}
+
+	out := filepath.Join(os.TempDir(), "azd-coverage-changed-files.txt")
```
Comment (jongio) [inline cli/azd/magefile.go]: This writes to a fixed path (`os.TempDir()/azd-coverage-changed-files.txt`). Two concurrent `mage coverage:diff` or `mage coverage:pr` runs on the same machine will clobber each other's file list, producing wrong gate results silently.

```suggestion
f, err := os.CreateTemp("", "azd-coverage-changed-files-*.txt")
if err != nil {
    return "", fmt.Errorf("creating changed-files temp file: %w", err)
}
out := f.Name()
f.Close()
```

**Review by jongio — COMMENTED** [review]
Solid rework - the two-gate model (per-package decrease + overall floor) with a single source of truth in `Get-CoverageDiff.ps1` is a clean design. wbreza's review covers the major items (AGENTS.md 65/69 mismatch, `-SkipPublisherCheck`, silent `os.RemoveAll` errors, missing Go tests); one additional item below.

**Review by jongio — APPROVED** [review]
Incremental review of the "address feedback" commit (1a85ef4).

My earlier comment about the fixed-path temp file race in `resolveChangedFilesForDiff` - addressed, now uses `os.CreateTemp`. The new test in `magefile_test.go` even verifies two calls produce distinct paths.

The pipeline changes are solid: baseline now pulls from the same platform matrix as the current build (via `DownloadArtifacts`), empty-dir detection is tighter (checks for actual files, not just directory existence), and the graceful degradation for missing baseline platforms is the right call - warns and continues instead of hard-failing.

Other fixes look good: Pester pinned to 5.7.1, mage pinned to v1.17.2, `os.RemoveAll` error handling, delta format shows "pp" instead of "%", docs aligned with implementation, unused `GH_TOKEN` removed.

Clean follow-up commit. All prior feedback addressed.
